amx/server/natives/a_players.lua

```diff
@@ -955,32 +1059,39 @@ end
 
 function EnableStuntBonusForAll(amx, enable)
 	notImplemented('EnableStuntBonusForAll')
+	return true
 end
 
 function EnableStuntBonusForPlayer(amx, player, enable)
 	notImplemented('EnableStuntBonusForPlayer')
+	return false
 end
 
 function TogglePlayerSpectating(amx, player, enable)
 	if enable then
 		fadeCamera(player, true)
 		setCameraMatrix(player, 75.461357116699, 64.600051879883, 51.685581207275, 149.75857543945, 131.53228759766, 40.597320556641)
+		toggleAllControls(player, false, true, false)
+		setPedWeaponSlot(player, 0)
+		setElementAlpha(player, 0)
+		setElementCollisionsEnabled(player, false)
 		setPlayerHudComponentVisible(player, 'radar', false)
 		setPlayerState(player, PLAYER_STATE_SPECTATING)
 	else
 		local playerdata = g_Players[getElemID(player)]
 		local spawninfo = playerdata.spawninfo or (g_PlayerClasses and g_PlayerClasses[playerdata.selectedclass])
-		if not spawninfo or playerdata.returntoclasssel then
+		if playerdata.returntoclasssel then
 			playerdata.returntoclasssel = nil
 			putPlayerInClassSelection(player)
-			return true
+		else
+			spawnPlayerBySelectedClass(player)
+			setPlayerHudComponentVisible(player, 'radar', true)
 		end
-		spawnPlayerBySelectedClass(player)
-		--In samp calling TogglePlayerSpectating also unsets camera interpolation
-		clientCall(player, 'removeCamHandlers')
+		setElementAlpha(player, 255)
```

Comment:
Up above we're disabling all controls when spectating, but never re-enabling them.
I also add `PLAYER_STATE_NONE` to mirror OpenMP's [behavior](https://github.com/openmultiplayer/open.mp/blob/master/Server/Source/player_impl.hpp#L637) to some degree.

```suggestion
		setElementAlpha(player, 255)
		toggleAllControls(player, true, true, false)
		setPlayerState(player, PLAYER_STATE_NONE)
```

Reply:
We're actually enabling all player controls when he is spawned unconditionally ([here](https://github.com/NexiusTailer/amx/blob/4abde183b23ef925eefb5cac1dfab943cbde697c/amx/server/events.lua#L280)), since SA-MP does it as well (you can freeze a player even directly with TogglePlayerControllable and he will be unfrozen after spawn anyway, so I replicated this behaviour). I can add some comment inside TogglePlayerSpectating to make it more clear because it can look like we really forget about it.

As for adding state none when leaving spectate like in omp, it is actually a critical behaviour difference with SA-MP server and I reported about it [there](https://github.com/openmultiplayer/open.mp/issues/1008) with reasons why. I suppose they've done it to bypass some other issue, but created a new one.

Reply:
> We're actually enabling all player controls when he is spawned unconditionally ([here](https://github.com/NexiusTailer/amx/blob/4abde183b23ef925eefb5cac1dfab943cbde697c/amx/server/events.lua#L280)), since SA-MP does it as well (you can freeze a player even directly with TogglePlayerControllable and he will be unfrozen after spawn anyway, so I replicated this behaviour). I can add some comment inside TogglePlayerSpectating to make it more clear because it can look like we really forgot about it.
> 
> As for setting state none when leaving spectate like in omp, it is actually a critical behaviour difference with SA-MP server and I reported about it [there](https://github.com/openmultiplayer/open.mp/issues/1008) with reasons why. I suppose they did it to bypass some other issue, but created a new one.

I agree on your points, but the current logic assumes we only spectate during one specific [stage](https://github.com/multitheftauto/mtasa-blue/blob/c992889b1b277f1bab4b602448676004e78ce8ca/Server/mods/deathmatch/logic/CMapManager.cpp#L620). It won't unfreeze the player or re-enable their controls if we for example say, behave as an administrator in a server and spectate a vehicle or player using `TogglePlayerSpectating`. When we're done spectating we will be frozen in place.